File: java-devildyw/src/main/java/top/dyw/leetcode/algorithm/贪心/L55跳跃游戏.java
package top.dyw.leetcode.algorithm.贪心;

/**
 * 给定一个非负整数数组 nums ，你最初位于数组的 第一个下标 。
 * <p>
 * 数组中的每个元素代表你在该位置可以跳跃的最大长度。
 * <p>
 * 判断你是否能够到达最后一个下标。
 *
 * @author Devil
 * @since 2022-10-20-13:57
 */
@SuppressWarnings("all")
public class L55跳跃游戏 {

    public static void main(String[] args) {

    }

    /**
     * 我们每次取最大的跳跃长度，这个就是可以条约的最大范围
     * 将问题转化为跳跃覆盖范围是否可以覆盖终点
     *
     *
     * @param nums
     * @return
     */
    public boolean canJump(int[] nums) {
        if (nums.length==1){
            return true;
        }
        //初始化覆盖范围为0
        int coverRange = 0;
        //在遍历数组的过程中更新最大的覆盖范围 （每次都取跳跃的最大长度）  coverRange = i+nums[i]
        //以覆盖范围的边界作为遍历的边界 也代表了跳跃的最大限度
        for (int i = 0; i <= coverRange; i++) {
            //更新覆盖范围
            coverRange = Math.max(coverRange,i+nums[i]);
            //如果覆盖范围可以到达最后一个下标
            if (coverRange>=nums.length-1){
                return true;
            }
        }

        return false;


    }

    public boolean canJump01(int[] nums) {
        if (nums.length==1){
            return true;
        }

        int coverRange = 0;

        for (int i = 0; i<=coverRange; i++){
            coverRange = Math.max(coverRange,i+nums[i]);
            if (coverRange>=nums.length-1){
                return true;
            }
        }

        return false;
    }
}

File: golang-devildyw/middle/L55跳跃游戏.go
package main

func main() {

}

// 贪心 我们关心到每一步能到达的最大距离，如果最大距离包含了最后节点 那么就能到达 反之不能
// 因为每一个下标对应数组的值的是当前能走的最大值 你也可以选择走1步 通过这个机制我们可以遍历数组 首先判断经过前面的更新 能够走到当前位置
// 如果能走到当前位置那么就更新能走到的最大位置

// 我们不需要知道“具体跳哪一步”，只需要知道“我目前能跳到的最远位置”是多少。
func canJump(nums []int) bool {
	var maxReach int = 0
	for i := 0; i < len(nums); i++ {
		if i > maxReach {
			return false // 到达不了 在第i个位置断档了 不可能到达最后位置了
		}
		maxReach = max(maxReach, i+nums[i]) // 如果能到达第i位置 那么更新能到达的最大位置
	}
	return true
}
